src/data/syncset/node.rs

```diff
@@ -0,0 +1,555 @@
+use std::cell::{Cell, RefCell};
+use std::mem;
+
+use crate::crypto::hash::{hash, Digest};
+use super::errors::*;
+use super::path::*;
+use super::Syncable;
+
+
+/// Private type used for the binary tree
+#[derive(Debug)]
+pub(super) enum Node<Data: Syncable> {
+    // Empty leaf
+    Empty,
+
+    // Non-empty leaf
+    Leaf {
+        // Data contained in the leaf
+        item: Data,
+        // Potentially empty cached hash
+        cached_label: RefCell<Option<Digest>>,
+    },
+
+    Internal {
+        // Pointer to the child nodes
+        right: Box<Node<Data>>,
+        left: Box<Node<Data>>,
+
+        // Pre-computed values for label and size
+        // todo: rename to label (+function)
+        cached_label: RefCell<Option<Digest>>,
+        cached_size: Cell<Option<usize>>,
+    },
+}
+
+impl<Data: Syncable> Node<Data> {
+    // todo? add node_at_mut? (Node is a private item structure, thus all uses would
+    // have to be within the syncset implementation)
+    /// Finds the first node at a given path. If a (potentially empty) Leaf node is encountered
+    /// prior to the path's max depth, a reference to that node is returned.
+    /// Otherwise, if the end of the path is reached, then then the iterated node will be returned
+    /// by reference.
+    pub fn node_at(&self, prefix: &Prefix, depth: usize) -> &Node<Data> {
+        if let Some(dir) = prefix.at(depth) {
+            if let Node::Internal { left, right, .. } = &self {
+                // Fork -> recurse into left or right
+                if dir == Direction::Left {
+                    left.node_at(prefix, depth + 1)
+                } else {
+                    right.node_at(prefix, depth + 1)
+                }
+            } else {
+                // Leaf -> this is the node wanted
+                &self
+            }
+        } else {
+            // End of path reached, this is the node wanted
+            &self
+        }
+    }
+
+    pub fn dump(&self) -> Vec<&Data> {
+        let mut result = Vec::with_capacity(self.size());
+        self.dump_recursive(&mut result);
+        debug_assert_eq!(result.len(), self.size());
+        result
+    }
+
+    fn dump_recursive<'a>(&'a self, result: &mut Vec<&'a Data>) {
+        match self {
+            Node::Leaf { item, .. } => result.push(item),
+            Node::Empty => (),
+            Node::Internal { left, right, ..} => {
+                left.dump_recursive(result);
+                right.dump_recursive(result);
+            }
+        }
+    }
+
+    /// Traverses the graph in a depth first manner (priority to left leaves), and applies the
+    /// function to each element encountered.
+    pub fn traverse<F>(&self, f: &mut F)
```

Comment:
According to rustc this function is never used, is it really needed?

Reply:
I'm not sure, it was in the original C++ implementation, and I wanted to preserve the interface (so I added it to `SyncSet` as well).

Reply:
So after discussing it with Matteo looks like it can just be removed :)